src/input/tablet.c

```diff
@@ -63,27 +66,97 @@ adjust_for_rotation(enum rotation rotation, double *x, double *y)
 	}
 }
 
+static struct wlr_surface*
+tablet_get_coords(struct drawing_tablet *tablet, double x, double y,
+		double *nx, double *ny)
+{
+	/* Convert coordinates: first [0, 1] => layout, then layout => surface */
+	double lx, ly;
+	wlr_cursor_absolute_to_layout_coords(tablet->seat->cursor, tablet->base,
+		x, y, &lx, &ly);
+
+	struct wlr_scene_node *node =
+		wlr_scene_node_at(&tablet->seat->server->scene->tree.node, lx, ly, nx, ny);
+
+	/* Find the surface and return it if it accepts tablet events */
+	struct wlr_surface *surface = lab_wlr_surface_from_node(node);
+
+	if (surface && !wlr_surface_accepts_tablet_v2(tablet->tablet_v2, surface)) {
+		return NULL;
+	}
+	return surface;
+}
+
 static void
 handle_axis(struct wl_listener *listener, void *data)
 {
 	struct wlr_tablet_tool_axis_event *ev = data;
 	struct drawing_tablet *tablet = ev->tablet->data;
+
 	if (ev->updated_axes & (WLR_TABLET_TOOL_AXIS_X | WLR_TABLET_TOOL_AXIS_Y)) {
 		if (ev->updated_axes & WLR_TABLET_TOOL_AXIS_X) {
 			tablet->x = ev->x;
 		}
 		if (ev->updated_axes & WLR_TABLET_TOOL_AXIS_Y) {
 			tablet->y = ev->y;
 		}
+	}
 
-		double x = tablet->x;
-		double y = tablet->y;
-		adjust_for_tablet_area(tablet->tablet->width_mm, tablet->tablet->height_mm,
-			rc.tablet.box, &x, &y);
-		adjust_for_rotation(rc.tablet.rotation, &x, &y);
-		cursor_emulate_move_absolute(tablet->seat, &ev->tablet->base, x, y, ev->time_msec);
+	double x = tablet->x;
+	double y = tablet->y;
+	adjust_for_tablet_area(tablet->tablet->width_mm, tablet->tablet->height_mm,
+		rc.tablet.box, &x, &y);
+	adjust_for_rotation(rc.tablet.rotation, &x, &y);
+
+	double nx, ny;
+	struct wlr_surface *surface = tablet_get_coords(tablet, x, y, &nx, &ny);
+
+	if (surface && tablet->tablet_v2_tool) {
+		wlr_tablet_v2_tablet_tool_notify_proximity_in(
+			tablet->tablet_v2_tool, tablet->tablet_v2, surface);
+
+		if (ev->updated_axes & (WLR_TABLET_TOOL_AXIS_X | WLR_TABLET_TOOL_AXIS_Y)) {
+			// Visualize where we are and what we are
+			wlr_cursor_warp_absolute(tablet->seat->cursor, tablet->base, x, y);
+			cursor_set(tablet->seat, LAB_CURSOR_PENCIL);
```

Comment:
Most likely not the best idea ..

Reply:
I don't think we actually need to set the cursor manually here at all.
`struct wlr_tablet_v2_tablet_tool` has a `set_cursor()` signal though that supplies a wl_surface, so I assume we can handle it similar to `request_cursor_notify()` in `src/input/cursor.c`.